test/Dialect/LLHD/Transforms/mem2reg.mlir

```diff
@@ -645,5 +645,132 @@ hw.module @DelayedConditionalDrives(in %u: i42, in %v: i42, in %w: i42, in %q: i
+    // CHECK-NEXT: [[TMP:%.+]] = hw.array_get %u[%i]
+    // CHECK-NEXT: [[FIELD1:%.+]] = comb.mux %q, %v, [[TMP]]
+    // CHECK-NEXT: [[DRV1:%.+]] = hw.array_inject %u[%i], [[FIELD1]]
+    llhd.drv %1, %v after %0 if %q : !hw.inout<i42>
+    // CHECK-NEXT: [[FIELD2:%.+]] = comb.mux %r, %w, [[FIELD1]]
+    // CHECK-NEXT: [[DRV2:%.+]] = hw.array_inject [[DRV1]][%i], [[FIELD2]]
+    llhd.drv %1, %w after %0 if %r : !hw.inout<i42>
+    // CHECK-NEXT: call @use_array_i42([[DRV2]])
+    // CHECK-NEXT: call @use_i42([[FIELD2]])
+    %2 = llhd.prb %a : !hw.inout<array<4xi42>>
+    %3 = llhd.prb %1 : !hw.inout<i42>
+    func.call @use_array_i42(%2) : (!hw.array<4xi42>) -> ()
+    func.call @use_i42(%3) : (i42) -> ()
+    // CHECK-NEXT: llhd.constant_time
+    // CHECK-NEXT: llhd.drv %a, [[DRV2]] after {{%.+}} :
+    // CHECK-NEXT: llhd.halt
+    llhd.halt
+  }
+  // CHECK: llhd.process
+  llhd.process {
+    // CHECK-NEXT: [[A:%.+]] = llhd.prb %a
+    // CHECK-NOT: llhd.sig.array_get
+    %1 = llhd.sig.array_get %a[%i] : !hw.inout<array<4xi42>>
+    // CHECK-NEXT: [[DRV1:%.+]] = comb.mux %q, %u, [[A]]
+    llhd.drv %a, %u after %0 if %q : !hw.inout<array<4xi42>>
+    // CHECK-NEXT: [[TMP:%.+]] = hw.array_get [[DRV1]][%i]
+    // CHECK-NEXT: [[FIELD2:%.+]] = comb.mux %r, %v, [[TMP]]
+    // CHECK-NEXT: [[DRV2:%.+]] = hw.array_inject [[DRV1]][%i], [[FIELD2]]
+    // CHECK-NEXT: [[ENABLE2:%.+]] = comb.or %r, %q
+    llhd.drv %1, %v after %0 if %r : !hw.inout<i42>
+    // CHECK-NEXT: [[FIELD3:%.+]] = comb.mux %s, %w, [[FIELD2]]
+    // CHECK-NEXT: [[DRV3:%.+]] = hw.array_inject [[DRV2]][%i], [[FIELD3]]
+    // CHECK-NEXT: [[ENABLE3:%.+]] = comb.or %s, [[ENABLE2]]
+    llhd.drv %1, %w after %0 if %s : !hw.inout<i42>
+    // CHECK-NEXT: call @use_array_i42([[DRV3]])
+    // CHECK-NEXT: call @use_i42([[FIELD3]])
+    %2 = llhd.prb %a : !hw.inout<array<4xi42>>
+    %3 = llhd.prb %1 : !hw.inout<i42>
+    func.call @use_array_i42(%2) : (!hw.array<4xi42>) -> ()
+    func.call @use_i42(%3) : (i42) -> ()
+    // CHECK-NEXT: llhd.constant_time
+    // CHECK-NEXT: llhd.drv %a, [[DRV3]] after {{%.+}} if [[ENABLE3]]
+    // CHECK-NEXT: llhd.halt
+    llhd.halt
+  }
+}
+
+// Delayed drives of signal projections should fall back to probing the current
+// value of the entire slot and injecting into that. This is equivalent to
+// treating all arrays as packed.
+// CHECK-LABEL: @DelayedProjectionDrive
+hw.module @DelayedProjectionDrive(in %u: !hw.array<4xi42>, in %v: i42, in %i: i2) {
+  %0 = llhd.constant_time <0ns, 1d, 0e>
+  %a = llhd.sig %u : !hw.array<4xi42>
+  // CHECK: llhd.process
+  llhd.process {
+    // CHECK-NEXT: [[A:%.+]] = llhd.prb %a
+    %1 = llhd.sig.array_get %a[%i] : !hw.inout<array<4xi42>>
+    // CHECK-NEXT: [[TMP:%.+]] = hw.array_inject [[A]][%i], %v
+    llhd.drv %1, %v after %0 : !hw.inout<i42>
+    // CHECK-NEXT: llhd.constant_time
+    // CHECK-NEXT: llhd.drv %a, [[TMP]] after {{%.+}}
+    // CHECK-NEXT: llhd.halt
+    llhd.halt
+  }
+}
+
```

Comment:
Maybe add tests for the following scenarios:
* Projected signal value is passed via block argument within a process (make sure it doesn't crash)
* Multiple projection ops refer to the same part of the original signal (make sure multiple drives to these different projections that refer to the same part update to the correct value)